### ISSUE TITLE
user.js breaks styling of right click menus in about:config, about:addons. Also breaks styling of Tree Style Tab

### ISSUE THREAD

**Issue by VoidSignal** [ghacksuserjs/ghacks-user.js#682]
![menu](https://user-images.githubusercontent.com/20287838/55422029-5cf2bf00-5583-11e9-9945-74824b17abb2.png)
On the left is a fresh profile with no user.js whatsoever, on the right is a fresh profile with the latest user.js from the master. Notice how on the right Tree Style Tab appearance also doesn't match system style.

I've read through the entire user.js twice, but I have no idea what might be causing it.

* Platform: Linux x86_64
* Firefox: 66.0.2
* Add-ons: None, except the Tree Style Tab to demonstrate the issue (reproducible without it)

**Comment by Thorin-Oakenpants** [comment]
I can't replicate in Windows. I will check in one of my Linux VMs later.

What theme are you using? Light/Dark? Custom? (might help me when I do tests).

About the only thing that can do this is RFP (pref 4501) and controlling system colors etc. Test if setting `privacy.resistFingerprinting` = false (and comment that line out in the user.js for this test, and restart to make sure) makes a difference.

PS: there are currently a couple of bugzillas open about this kind of thing - i.e FP'ing colors - so I'm almost 100% sure this is the cause

**Comment by VoidSignal** [comment]
I'm using Arc Dark theme (the latest version from Arch Linux repositories). Disabling `privacy.resistFingerprinting` didn't solve the problem.

**Comment by Thorin-Oakenpants** [comment]
I can't find TST on AMO: https://addons.mozilla.org/en-US/firefox/search/?q=tree style tab&sort=users&type=extension  - all I see are similar names and nothing like that name with any decent numbers. Probably not necessary since issue also shows in about:config context menu

Not that the about:config changes in 67: there is no longer a context menu

So I installed Arc Dark theme in a nilla profile in Windows: the version is 6 months old (2018.9.24) and couldn't replicate (but I only tried with a few settings). I am not very good with Linux. Can you provide a link to the Arc Dark theme at the Arch Linux repos (no sure how I would even load this, and I only have a template Arch VM with no desktop environment: so I will test on Ubuntu/Debian/Mint first).

While I think about this and have some food and stuff ... try resetting, commenting out, and restarting the browser with `ui.use_standins_for_native_colors` (hidden pref) removed

**Comment by VoidSignal** [comment]
Here's [Tree Style Tab](https://addons.mozilla.org/en-US/firefox/addon/tree-style-tab/) and here's [Arc theme](https://www.archlinux.org/packages/community/any/arc-gtk-theme). Click on "Download From Mirror" and extract the contents of `/usr/share/themes/` to wherever themes go on Ubuntu. I'd guess they go exactly into `/usr/share/themes/`.

But there's no need, because disabling `ui.use_standins_for_native_colors` fixes the issue. I guess what remains how is to warn others of possible breakage?

**Comment by Thorin-Oakenpants** [comment]
> But there's no need

Cool. Because I don't think I would be able to replicate

> I guess what remains how is to warn others of possible breakage

Well... we'd probably have to establish under what situations this happens - it might only be Linux and GTK stuff, IDK (am not an expert) ... **but**, that pref is going to be moved to RFP ALTs section and disabled by default come FF67, in about 4 weeks time. But then those not using RFP would/should enable it. I can add a general `[SETUP` tag to it : "may break some themes"

from my sticky reminder
* [1485266](https://bugzilla.mozilla.org/show_bug.cgi?id=1485266) - RFP ALTS <- disable exposure of system colors to CSS or canvas (see 2618) (FF67+)

I will say that there are other bugzillas that may change this behavior - such as not applying RFP to extensions, to removing GTK in Linux builds, and more - at least four issues

Thanks for the troubleshooting :kiss:

**Comment by Thorin-Oakenpants** [comment]
Actually .. and I'm not sure, guess you will find out in FF67+ ... if I read that ticket correctly, RFP will re-introduce the problem for you

If it does, then let us know

**Comment by VoidSignal** [comment]
Okay, I'll keep you posted.

Why are they even shoveling so much into a single pref?

**Comment by Thorin-Oakenpants** [comment]
Without looking, it might even come down to fixing the issue within the theme/extension. If they used there own css rules instead of the system ones, the problem wouldn't happen, if I understand it correctly. Although using the preset ones means less overhead and is always in line with FF's dark/light theme etc. Not my area of expertise, sorry. So if it does happen in 67+, your best best is to address it with the theme/ext devs - or it's a bug, IDK.

> Why are they even shoveling so much into a single pref

It's a master switch, and anti-fingerprinting covers sooooooo much stuff. It touches almost every aspect of the browser and web tech. From a coding perspective, it's easier to call a single reference to check if it's on or off, I guess.

**Comment by Thorin-Oakenpants** [comment]
@VoidSignal 

Just FYI : https://thorin-oakenpants.github.io/testing/ (while it's there) - you can see your system color fingerprint.
- `ui.use_standins_for_native_colors` mitigates this
- RFP in FF67 mitigates this

Here is the "protected" hash as per TB, or when RFP or that pref is used
- `b5e2344c265fc498d2fb8e0f84951e8d501ad481`

Here are some results (with no mitigation)
```
- 
- TB etc:       b5e2344c265fc498d2fb8e0f84951e8d501ad481
- Win7 default: b5e2344c265fc498d2fb8e0f84951e8d501ad481 - default (same as TB)
- Win7 custom:  e26af9c700cd46c6528f945febe722d365a23087 - changed desktop background color
- Mint:         7eebf18db35b444aa1ac179f2d152b8d8096ba37 - some results have an alpha value (rgba)
- Ubuntu:       aa8743c28ed4c20f5505cb840de4115da0b95ef1 - some results have an alpha value (rgba)
- Debian:       93e3efd6c7d01008884e939fd77ece10a836cc29 - no alpha values
- openSUSE:     b46ecf280012aac7183e5f565965ba30e319090d - no alpha values
```

I played with themes in Win7 - what confused me for a while was that the default Win7 setup was the same hash as the TB/protected one. And changing Win7 themes from default to one of the `fancy named built in ones` (like "Architecture") doesn't change the system colors - to do that you have to apply the ones that relate to high contrast etc. And of course if you actually change any system colors manually, e.g I set my VMs up with a solid custom color background, then that will change the hash. You also need to restart FF between theme changes.

I didn't play with themes in Linux. But I assume the same would apply.

Also, my android gives yet another hash (when not protected). But I can't be assed testing any more.

So, FF66 or lower, the ui.use_standins* pref protects you, and in FF67+ RFP protects you - all tested. It's definitely fairly high entropy IMO - more so than detecting e.g. Linux flavors, as who doesn't customize their setups. That said, I highly doubt anyone is using it to FP. It's just one of those things that need to be covered to reduce the attack surface.

Long story short, I am 99% sure your original problem re styling in extensions/themes will resurface in 67.

**Comment by VoidSignal** [comment]
@Thorin-Oakenpants thanks for the follow up.

>I didn't play with themes in Linux. But I assume the same would apply.

I've run the test and got a different hash from everything you've listed, so you are indeed correct.

The original issue is quickly becoming irrelevant for me since I now set `user_pref("widget.content.gtk-theme-override", "Arc")` to enforce light variant of the system theme for page elements. This pref also affects the default appearance of Tree Style Tab which now of course matches Arc theme, not Arc Dark. In any case, I'll have to style it manually to fit it with the rest of the interface. Fortunately, there's an option for that.

At this point, I've come to the conclusion that it is an issue with Firefox itself. I have no idea why it uses system colors on web pages in the first place instead of shipping sensible cross-platform defaults. The fact that Tree Style Tab blends so nicely with the rest of the interface is the only nice consequence of that.

**Comment by Thorin-Oakenpants** [comment]
> I have no idea why it uses system colors on web pages

Can you give me an element on a page that happens on? Because I'm trying to work out if theme colors can be read by content = FP'able. See https://bugzilla.mozilla.org/show_bug.cgi?id=1540726, which I'm not sure what to make off. It certainly sounds like from the comments, that they can. I'm guessing all I need to do is just look up all the FF theme css rules and build a test

**Comment by Thorin-Oakenpants** [comment]
PS: will edit your top post with your solution

**Comment by VoidSignal** [comment]
>PS: will edit your top post with your solution

This is not a solution. `widget.content.gtk-theme-override` does not fix context menus in about:config and about:addons and it doesn't return system styling to Tree Style Tab.

From now on let "gtk pref" = `widget.content.gtk-theme-override` and "colors pref" = `ui.use_standins_for_native_colors`.

Firefox applies system style to "widgets" (checkboxes, input fields, buttons, dropdown lists...). This is their default style that can be overridden in CSS. Gtk pref merely provides a way to change what theme is used. Widget style **is not** affected by colors pref and only obeys gtk pref.

Gtk pref is needed because when you use dark theme widgets often don't fit with the rest of the page when defaults are used. You also might get black text on dark background if for a certain widget text color is explicitly set to black, but background color is not set.

I'm assuming there's a JS API to read system colors, otherwise you couldn't have created that test. Gtk pref also affects what values are supplied via that API. I get vastly different results on your test when gtk pref is not set (Arc Dark is used), when it's set to Ark (Ark is used) and when colors pref is on (some spoofed values are used).

Apparently, Tree Style Tab uses that API to blend with the rest of the interface. If colors pref is on, of course it can't do that and that was what I observed initially.

Later I turned color pref off as you suggested, but I also set gtk pref to Ark for the reasons outlined above. That caused Tree Style Tab to use colors from Ark, but I wanted it to use colors from Ark Dark which is impossible unless I unset gtk pref which is undesirable. That's what I meant when I said that the issue is no longer relevant. If color pref is on, Tree Style Tab uses spoofed colors. If color pref is off it uses Ark colors. Neither is what I want and I'll have to style it manually anyway.

Here are some pics:

* uMatrix settings

![umatrix](https://user-images.githubusercontent.com/20287838/55918866-48e23980-5be3-11e9-99af-4f5ad69eb8ae.png)

* privacytools.io subreddit

![reddit](https://user-images.githubusercontent.com/20287838/55918969-b42c0b80-5be3-11e9-938a-4c33fed09b5f.png)

You can clearly see that they have Arc Dark styling. If I set `user_pref("widget.content.gtk-theme-override", "Arc")` they will be styled accordingly.

**Comment by Thorin-Oakenpants** [comment]
> This is not a solution

OK, I removed the edit in OP

---

re my question about themes being leaked, I worked it out. It's a FF67+ standard
- https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme
- Only Safari (nov? last year) and FF (now) have this
- allows websites to create their own color schemes and use them against a returned value of light, dark, or no-preference
- RFP enforces "light" - meaning when this becomes more mainstream (e.g chrome add it), it'll mean you get served light schemes on websites that use it regardless of your theme
- this has no bearing on your issue

---

GTK is most likely going to be removed, AFAIK

"I'm assuming there's a JS API" - just normal css and js. I set an element's `background-color` to e.g `ActiveBorder` and then read back the background color property .. and repeat

I'm not sure what the solution is for you. Do themes have to be signed? If not you could probably just change the css rules for widgets and reload it?

**Comment by VoidSignal** [comment]
>GTK is most likely going to be removed, AFAIK

Do you mean `widget.content.gtk-theme-override` will be removed soon?

>Do themes have to be signed? If not you could probably just change the css rules for widgets and reload it?

I'm not sure what you mean by that.

**Comment by Thorin-Oakenpants** [comment]
re GTK: and is this from memory, from something I read, ... I **think** I read that they **proposed** removing it entirely from Firefox so as to unify the code across platforms - if they do, it's nothing to worry about. It would remove issues like you experienced in this topic - as well as other ones that could help FP you.

re: signed: i mean that extensions must be signed (and Firefox stable will not let you install unsigned extensions) - this is a safety measure so rogue malicious extensions can't be sideloaded. If you modified an extension (opened the xpi, changed it), then Firefox will not use it because it's signature has changed. etc.

But **themes** can't be malicious, so if they're not signed you could modify the theme to fix those widgets.  Or just use a userChrome.css <-- which is probably the easiest fix

**Comment by Thorin-Oakenpants** [comment]
^^ here you go. I guess I read too much stuff, so no idea why this was in my memory - 
https://bugzilla.mozilla.org/show_bug.cgi?id=1411425 - it's just a proposal started 2 years ago

**Comment by Atavic** [comment]
Just read the 1st post, and it seems a DOM issueto me.
Assuming you right-clicked on the about:config entry seen in the image, I point you to:

*2401: disable website control over browser right-click context menu*

* 2710: disable DOM (Document Object Model) Storage* and its *[WARNING]*

... on the user.js

**Comment by Atavic** [comment]
By looking at the 1st post, it seems a DOM issueto me.
Assuming you right-clicked on the about:config entry seen in the image, I point you to:

*2401: disable website control over browser right-click context menu*

*2710: disable DOM (Document Object Model) Storage* and its *[WARNING]*

... on the user.js

**Comment by Thorin-Oakenpants** [comment]
@Atavic OP isn't worried about those. Right click in OP is not on a web content page (and if it was, see shift-right-click). There's no issue here getting a context menu. Not sure what dom storage has to do with this - the TST example is working (by the looks of it) and being on FF66 I assume extension storage->IDB. Both prefs are inactive by default in the user.js

The issue is just about styling.
